packages/protocol/contracts/identity/FederatedAttestations.sol

```diff
@@ -0,0 +1,386 @@
+    } else {
+      return (accounts, issuedOns, signers);
+    }
+  }
+
+  /**
+   * @notice Returns identifiers mapped (by unrevoked signers) to a
+   *   given account address by a list of trusted issuers
+   * @param account Address of the account
+   * @param trustedIssuers Array of n issuers whose identifier mappings will be used
+   * @param maxIdentifiers Limit the number of identifiers that will be returned
+   * @return Array (length <= maxIdentifiers) of identifiers
+   * @dev Adds identifier info to the arrays in order of provided trustedIssuers
+   * @dev Expectation that only one attestation exists per (identifier, issuer, account)
+   */
+
+  // TODO ASv2 consider also returning an array with counts of identifiers per issuer
+
+  function lookupIdentifiersByAddress(
+    address account,
+    address[] memory trustedIssuers,
+    uint256 maxIdentifiers
+  ) public view returns (bytes32[] memory) {
+    // Same as for the other lookup, preallocate and then trim for now
+    uint256 currIndex = 0;
+    bytes32[] memory identifiers = new bytes32[](maxIdentifiers);
+
+    for (uint256 i = 0; i < trustedIssuers.length; i = i.add(1)) {
+      address trustedIssuer = trustedIssuers[i];
+      uint256 numIssuersForAddress = addressToIdentifiers[account][trustedIssuer].length;
+      for (uint256 j = 0; j < numIssuersForAddress; j = j.add(1)) {
+        // Iterate through the list of identifiers
+        if (currIndex < maxIdentifiers) {
+          bytes32 identifier = addressToIdentifiers[account][trustedIssuer][j];
+          // Check if the mapping was produced by a revoked signer
+          // solhint-disable-next-line max-line-length
+          OwnershipAttestation[] memory attestationsForIssuer = identifierToAddresses[identifier][trustedIssuer];
+          for (uint256 k = 0; k < attestationsForIssuer.length; k = k.add(1)) {
+            OwnershipAttestation memory attestation = attestationsForIssuer[k];
+            // (identifier, account, issuer) tuples should be unique
+            if (attestation.account == account && !revokedSigners[attestation.signer]) {
+              identifiers[currIndex] = identifier;
+              currIndex = currIndex.add(1);
+              break;
+            }
+          }
+        } else {
+          break;
+        }
+      }
+    }
+    if (currIndex < maxIdentifiers) {
+      // Allocate and fill properly-sized array
+      bytes32[] memory trimmedIdentifiers = new bytes32[](currIndex);
+      for (uint256 i = 0; i < currIndex; i = i.add(1)) {
+        trimmedIdentifiers[i] = identifiers[i];
+      }
+      return trimmedIdentifiers;
+    } else {
+      return identifiers;
+    }
+  }
+
+  // TODO do we want to restrict permissions, or should anyone
+  // with a valid signature be able to register an attestation?
+  modifier isValidUser(address issuer, address account) {
+    require(
+      msg.sender == account ||
+        msg.sender == issuer ||
```

Comment:
My understanding here is that the `issuer` is the parent `account` and not the `signer`. The parent `account` should not be a hot wallet since this needs to be kept most secure. The `signer` would be better, but I'm still leaning toward the idea of letting anyone add a valid attestation as long as it can be marked as `revoked` by the user or issuer.

Reply:
Agreed with encouraging issuers to be generally unused; currently this should also allows signers to upload (line after this one, currently 247).

We aren't currently allowing individual attestations to be revoked; that said, per the discussion on Notion, this could be mitigated by only allowing upload during a small window around the publishedOn date. In that case, deleting the attestation in most cases ends up being pretty close to revocation, which I think is actually ok for the types of attestations we are targeting (impermanent, lightweight, PN/social).

Maybe this is good evidence for 1) moving towards a publishedOn + time window scheme instead of issuedOn 2) allowing upload by anyone!

cc @isabellewei

Reply:
Can you point me to the Notion discussion around not allowing individual attestations to be revoked?